### DISCUSSION TITLE
Power lambdas (suggestions for a better name welcome)

### DISCUSSION THREAD

**Discussion by jgonggrijp** [mustache/spec#155] [Ideas]
This is a continuation of the discussion in #135. Before that, there was also relevant discussion in #19 and #41, including a comment that hinted at something like power lambdas in https://github.com/mustache/spec/issues/41#issuecomment-7337582.

At the time of writing, (regular) lambdas have the following *required* capabilities, according to the interpolation and sections specs:

- If used in a section, receive the template content of the section as a string argument.
- Return a value that is rendered (for interpolation) or used as the new context (for sections), instead of the lambda (function/method) itself.

The *optional* lambdas extension adds one more capability:

- If the return value of the lambda is a string, render it as a template. In a section, this template replaces the original content of the section.

Power lambdas would supercharge the capabilities of lambdas, while changing as little about the language as possible. There is no syntactic change and only one semantic change: lambdas that can handle it receive an additional argument, which I refer to as the "magic".

The magic adds one or more of the following features to the existing capabilities of lambdas, in an implementation-defined way. Implementations are free to omit some of them, but encouraged to support as many as they can. **However, it is up to debate whether all of the following features should be included in the spec.** More on this below.

*Edit: I no longer believe features 1-3 are needed, see https://github.com/mustache/spec/discussions/155#discussioncomment-7482941.*

1. Retrieve a list of all keys that are visible in the context stack.
2. Look up the value of a given key in the context stack. Ideally, it must also be possible to see which context frame underlies the value that was found.
3. Retrieve any frame in the context stack (intended as read only).
4. Override the content with an external template that was already compiled beforehand (rather than returning a string that will be compiled or interpreted as a template on the fly).
5. Override the content with a final string that is not processed further (rather than returning a string that will be compiled or interpreted as a template). For sections, this string will not be escaped; if the string needs to be escaped, the power lambda should do it.
6. Render a template of choice against the current context stack and obtain the result as a string, without it necessarily being interpolated.
7. Define a prefix and/or suffix, in order to wrap whatever ends up being rendered in the place of the interpolation tag or section. This can currently already be done in section lambdas by concatenating the prefix, section content argument and prefix and returning the resulting string as a new template, but this feature makes it possible for interpolation tags as well and removes the necessity to compile or interpret a template at runtime. Implementations may support static strings, template strings or precompiled templates as prefix/suffix, or at their option, multiple of those.

I envision this as an optional extension module, perhaps `~power-lambdas.yml`. It would follow a slightly different format from the existing `~lambdas.yml`: each implementation that attempts to pass the specs in that module will need to bring its own lambdas.

The first three features mostly cater to a lambda's ability to return a value that gets pushed on the context stack. The other four mostly cater to a lambda's ability to override the template inside a section. There is some interaction between these sets as well. In detail, the features were originally motivated as follows:

- 1+2 enable advanced masking. For example, a lambda might return a new context frame in which every list in the context is masked by (a lambda proxy to) an enumerated version of that list. Such masking would make lambdas capable of everything that filters could do (see #153) and more. Where a filter would be written `{{# someList | enumerate }}`, the equivalent power lambda would be written `{{# enumerate.someList }}`.
- 2 alone would suffice to implement a more limited alternative to filters that only works for interpolation. Instead of `{{ someString | capitalize }}`, one would be able to write `{{#capitalize}}someString{{/capitalize}}`. However, passing section content as argument text to lambdas in this way would also open some options that filters do not have, since the lambda is free to interpret the section content in any way that suits the application. Feature 5 could make this use case more efficient.
- 1+2+3 OR 1+(an enhanced version of 2) would make it possible to implement skips (see #154) without adding any new syntax to the Mustache language.
- 3 alone would make it possible to implement anchored descent as well as root anchoring, as discussed in #154.
- 2+4 OR 2+5+6 would enable an efficient lambda-based implementation of dynamic partials. Instead of `{{>*animal}}` one would write `{{#dynPartial}}animal{{/dynPartial}}` (for example). 2 alone would also suffice, but would require recompiling the external template every time. *This use case is no longer relevant because dynamic names already made it into the spec in the meanwhile.*
- 4 OR 5+6 OR 7 would make it possible for a lambda to override or extend a local template, without that template having to be compiled at runtime. This would be helpful for implementations that are safely typed, such as JStachio by @agentgt.
- 4 OR 7 would also make it possible for a lambda to override or extend a local template *and* push a new frame on the context stack at the same time.
- 5 alone would restore a capability that was lost with the lambdas extension: to return a string that will be interpolated verbatim (or used as context frame), without it being subject to expensive template processing.
- 5+6 would enable an inversion of control, where the lambda renders a template and returns a result that is interpolated, instead of it returning a template that is then rendered and interpolated. Together with the existing capability to read the content of a section, this would enable me to write a Mustache equivalent of [handlebars-i18next](https://www.npmjs.com/package/handlebars-i18next).
- More generally, 6 alone would enable lambdas to derive information from the context. In this sense, it could also be considered a weak alternative to feature 2.
- 7 alone would provide an optimization for a use case that is often seen in example code: take the content of a section, add a prefix or suffix to it and return the result as a template override.

For the first two motivations, where power lambdas provide a close alternative to filters, I now believe that filters are more appropriate, even though power lambdas would be more powerful. Filters are much easier to implement for a given application: a filter can be any straightforward function in the host language that maps one value to another, while a power lambda requires specific knowledge about the Mustache engine as well as intricate masking and/or argument parsing logic. In addition, I cannot think of a convincing situation where a filter-like use case really requires the additional context awareness or argument parsing that a power lambda could provide.

The next two motivations, which position power lambdas as a solution for the masking problem, are more convincing in my opinion. Syntactic alternatives are discussed in #154, but I feel none of those alternatives is very good, so I currently still have a slight preference for the lambda approach. That said, if a syntactic (or other non-lambda) solution for the masking problem does make it into the spec, this completely removes any justification for feature 3. *Update: I think I have now come up with a very good alternative, see https://github.com/mustache/spec/discussions/155#discussioncomment-7482941. Features 1-3 can all be left out if that alternative is implemented.*

Dynamic names are in the spec already. I personally find the remaining motivations very convincing, but since those only involve features 4-7, this makes the inclusion of features 1-3 a bit uncertain as I see it. Of the latter three, 2 is probably the most useful one, but also the most redundant with filters. Still, there might be other good arguments why 1-3 need to be included anyway.

7 is nice to have, but not strictly necessary if 4-6 are also available. Thus, 4-6 seem to be the most "core" features of power lambdas, with 7 being a welcome extra.

An interesting option would be to return a lambda from a filter. That is, a filter might construct and return a closure, bound method or partially applied function with knowledge about a value in the context. That function could then proceed to act as a lambda, with context argument parsing, template overrides and/or pushing a new stack frame. I have discussed this option in #153 in more detail. It could provide a partial alternative to features 1-3, though in many cases, it may not be enough to address the masking problem.

**Comment by jgonggrijp** [comment]
In #135, @agentgt suggested investigating exising implementations of "more powerful" lambdas, in order to get an overview of existing practice. I replied with this list of implementations I was already aware of:

- Near the beginning of #135, @bobthecow described the power features in Mustache.php. If I understood correctly, he implemented at least feature 6 (render an arbitrary template against the current context stack and obtain the result as a string), but not feature 1 or 3 (full read access to the context stack).
- Mustache.php also implements filters (#153), though I would consider that just "different" from lambdas rather than "more advanced". I mention it for completeness.
- In https://github.com/mustache/spec/pull/147#issuecomment-1590249650, @spullara mentioned having "more powerful lambdas" in Mustache.java. He didn't clarify what exactly makes them more powerful, but I suppose that could be found in the documentation.

**Reply by jgonggrijp** [reply]
In https://github.com/mustache/spec/pull/147#issuecomment-1793791169, @agentgt briefly returned to this topic with the following comment:

> (...) I believe we were talking about searching for other Mustache implementations that have power(like) lambda support and JMustache does and JStachio was largely modeled after it with the exception of getting the whole stack (JStachio does only top of stack): https://www.javadoc.io/doc/com.samskivert/jmustache/latest/com/samskivert/mustache/Mustache.Lambda.html
> 
> We have been using a forked version of JMustache for almost a decade. It's small yet fairly powerful. In fact we tried switching to `Mustache.java` several times but found it too hard to lock down for our needs (we have quasi users editing templates) and I could not port some lambdas that we had as `Mustache.java` does not have stack access last time I checked. (...)

**Comment by jgonggrijp** [comment]
In https://github.com/mustache/spec/discussions/153#discussioncomment-7482911, I came up with a flavor of "power filter" that could address the masking problem. I think it provides a cleaner alternative to features 1-3. It is less powerful, but still powerful enough to make me want to leave features 1-3 out of the power lambdas specification!

**Comment by agentgt** [comment]
@jgonggrijp On discovery of other implementations: Another Mustache variant in Java land that has sort-of power lambdas is Trimou: https://github.com/trimou/trimou

Trimou went the Handlebars helper route. I'm not sure why Trimou never got that popular. My guess is people in Java land would either just accept the decorating or prebuilding of models (i18n) required in Mustache.java (@spullara) or use actual Handlebars.java. Or probably just marketing and network effect.

Several times I have considered experimenting with a Handlebars helpers like extension to JStachio but the only real use case I seem to need it for was i18n.

Sort of an open question but I have to wonder how much of power lambda desire / use case is i18n particularly parameters?

In the past when we did i18n in my company we used sort of a giant key/value (`Map<String,String>`) with multiple overlays (e.g. first try this map then try the other). There were some scaling issues but generally worked particularly if we used sessions.

In other cases the model would just get all of the i18n resolved apriori.

I guess it would be interesting to me and possibly helpful to others if we had a list of fairly concrete use cases power lambdas could solve.

Otherwise it appears where we have been using JStachio/JMustache lambdas a filter could have been used instead (well I'm not so sure on the whole fallback stuff but the transform this node to another node is mainly what we are using lambdas for).

**Reply by jgonggrijp** [reply]
First off, I agree that we currently end up "abusing" lambdas for things that filters can handle in a much simpler and more elegant way. Unless we use an implementation that already has filters, of course.

As for concrete examples, I think i18n is a good one. Here is roughly how I would approach it in JavaScript, based on my current ideas about how power lambdas and filters should work:

```
{{# friend | i18n }}{{#key}}goodmorning{{/key}}{{#t}}Good morning, {{name}}{{/t}}{{/ friend | i18n }}
```

```js
{
    // The friend is going to be our payload below.
    friend: {
        name: 'Ngozi'
    },
    // i18n is a filter that returns a lambda!
    i18n: function(payload) {
        // The payload is what the i18n framework needs for interpolation.
        // Thanks to the filter, our lambda below is aware of the payload.
        return function() {
            // The lambda pushes a frame on the stack with additional lambdas.
            // In this example, I create the object on the fly in order to keep
            // all code in one place, but in reality I would call a constructor.
            return {
                // member that will hold state
                keys: [],
                // key is a regular lambda that leaves no output.
                key: function(content) {
                    // We allow multiple keys, because the framework supports fallbacks.
                    this.keys.push(content);
                    // We return a falsy value to prevent the section from rendering.
                    return null;
                },
                // i18next supports other arguments besides the keys. I omit them
                // here for brevity.

                // t actually needs its power abilities, to first give the i18n framework
                // a chance at interpolation and fall back to Mustache interpolation
                // if necessary.
                t: function(content, magic) {
                    // We start by trying to use the framework.
                    var translation = i18next.t(this.keys, content, payload);
                    // Fallback case: key could not be matched. Push payload onto the
                    // stack and let Mustache render the default string.
                    if (!translation) return payload;
                    // Fallback case: framework was unable to interpolate all values.
                    // Let Mustache do the remaining ones; this requires overriding the
                    // template AND pushing the payload onto the stack.
                    if (/\{\{/.test(translation)) {
                        magic.replaceContentTemplate(translation);
                        return payload;
                    }
                    // Success case: translation is complete. We could just return it,
                    // but we will use the magic to prevent Mustache from needlessly
                    // compiling it.
                    magic.replaceContentStatic(translation);
                }
            }
        }
    }
}
```

Now, suppose we have the following i18next config and the current language is French.

```json
{
    "de": {
        "goodmorning": "Gute Morgen, {{name}}"
    },
    "fr": {
        "goodmorning": "Bonne matin {{name}}"
    }
}
```

Then the above template should produce `Bonne matin Ngozi`.

If you have other ideas for examples and would like to see how I might implement them, please let me know.

**Reply by spullara** [reply]
Huh. The twitter one in this case would have just looked more like:

```
{{#friend}}{{#i18n}}Good morning, {{name}}{{/#i18n}}{{/friend}}
```

We would pass a special i18n extraction handler over it to just extract the "Good morning, {{name}}" which would be sent to translation. There was another special case for countable things that I can't quite remember right now since some languages have up to 6 plural variants.

**Reply by jgonggrijp** [reply]
Right, so you have power lambdas with stack access. And you happened to use the default string as the only key. i18next allows that, too, but I wanted to illustrate both ways of passing information to a lambda.

**Reply by agentgt** [reply]
You do not need power lambdas. You just need the current existing optional lambda extension of returning a template. 

What @spullara was doing is basically the venerable GNU gettext. `Good morning, {{name}}` is key in a translation table which also happens to be Mustache template and translations would also be mustache templates.

Now if @spullara wants to know why we didn't do that is because unlike Twitter where there is probably only one template with

```
{{#friend}}{{#i18n}}Good morning, {{name}}{{/#i18n}}{{/friend}}
```

We have thousands of customers who have custom themes with i18n in them. Additionally each customer could customize the translation key/value table so a symbol as the key was better for our use case.

```
{{#friend}}{{i18n.GOOD_MORNING}}{{/friend}}
```

The `{{name}}` part was basically the personal session based translation table that would be interpolated and cached. A global one (e.g. anonymous users) would avoid that personalization so parameters were not needed.

The only tricky part was plurals and I can't recall what we did/do. I will find out later today.

Also lambda optional extension was not well supported in 2010 or I just didn't know about it.

Independent of the above JStachio struggles with @spullara technique because templates have to be known by the compiler in advance (each translation needing a pass of javac does not scale well). So in the meantime traditional Java i18n techniques (aka resource bundles) of positional parameters instead of mustache bindings have been used.

In the Handlebars-esque Java world (handlebars.java, trimou) they are following the more traditional Java i18n method of positional parameters.

The above techniques are error prone and I have many times explored how to make i18n more type safe but don't have a good solution yet.

**Reply by jgonggrijp** [reply]
> You do not need power lambdas. You just need the current existing optional lambda extension of returning a template.

By "you" do you mean to refer to me or to @spullara?

Adding to what you wrote, of course it is possible to implement i18n in a way that does not require power lambdas or filters. In fact, it could be done with no lambdas at all. In one project that used Mustache, I simply registered every translation string as a partial:

```
{{#friend}}{{>goodmorning}}{{/friend}}
```

The other notation that you showed amounts to an even simpler system, assuming that `i18n.GOOD_MORNING` directly resolves to a string rather than a lambda that returns a template:

```
{{#friend}}{{i18n.GOOD_MORNING}}{{/friend}}
```

As you said in other words, @spullara's notation is nearly the same except that the key also happens to be the default fallback:

```
{{#friend}}{{#i18n}}Good morning, {{name}}{{/#i18n}}{{/friend}}
```

However, all these approaches assume a low-tech, homegrown solution where the translation string is just a Mustache template. I am interfacing with [i18next](https://www.i18next.com/), which supports fallthrough multiple key lookup with defaults, as well as gender, number, namespaced translations, formatting, automatic language detection and many other bells and whistles. It also has an interpolation syntax that looks similar to Mustache but has different rules. Trying to interface with i18next using only old fashioned extension lambdas will not get you very far, most importantly because i18next will try to interpolate the variables and fail at that because you did not pass the necessary information.

In Handlebars, helpers can access the stack and I was able to make the following notation work by writing [this helper](https://github.com/UUDigitalHumanitieslab/handlebars-i18next):

```
{{#friend}}{{#i18n 'GOOD_MORNING'}}Good morning, {{name}}{{/#i18n}}{{/friend}}
```

A crucial difference from the previous examples is that the i18n helper will pass the `friend` object to i18next and that i18next will do the interpolation. I can also pass additional arguments to the helper, unlocking the entire API of `i18next.t` within the template.

Doing something like that in Mustache would require either power lambdas with stack access, or power lambdas without stack access but combined with filters. The latter scenario is what I was illustrating in my previous comment.

**Reply by agentgt** [reply]
> By "you" do you mean to refer to me or to @spullara?

Previously you (@jgonggrijp) said:

> Right, so you have power lambdas with stack access. 

In an indirect way I'm trying to clarify that Mustache.java lambdas does not have stack access unless you want to call stack access returning a template that will get interpolated. @spullara can correct me if I'm wrong.

There was similar confusion in #157 (EDIT actually in #147) with @spullara decorator (which I think is still you manually have to apply) is not a power lambda/filter.

You probably know this but I just want to keep it clear for others.

**Reply by spullara** [reply]
Mustache.java doesn't currently expose the context stack to lambdas (though TemplateFunctions get reinterpreted at runtime and may provide a similar benefit) though that isn't difficult to add if we agree to it. That said, I haven't had a need for it nor people asking for it outside this group. As for decorators in mustache.java, they can be added at a higher level so they are applied to all applicable objects in the context without having to do it explicitly on a collection by collection basis. Mustache.java is extremely customizable and many of the users of it have their own unique weird things that they do and I definitely wouldn't want all of them in the spec :)

**Reply by agentgt** [reply]
> their own unique weird things that they do and I definitely wouldn't want all of them in the spec :)

Yeah my point to highlight how we do i18n is to show how varied this is. i18n is going to vary and that i18next seems to be @jgonggrijp  major use case for power lambdas at the moment. I'm sure he has others and that is kind of what I want to see.

For example if I didn't have users editing templates in early 2010 I would have picked your (@spullara)  approach and I think given smart preprocessing (e.g. a better gettext) can actually be less error prone than i18next from what I have seen. That is bad key usage could be found at compile time (well preprocess time). 

In terms of leveraging Mustache syntax for i18n interpolation it is difficult for static JStachio but a simple solution is just to take the top of the stack which JStachio allows and pass it to JMustache or Mustache.java to do the dynamic template rendering. 

That is JStachio in some cases needs more powerful lambdas (which it does already have) than others  because it is less dynamic but I'm not sure the same can be said across other implementations.

On the other hand:

Because JStachio has virtual keys for lists one of the most often usages of lambdas is doing something to `@index`.

What my developers complained about is transformation fatigue and the off shore team will complain that  the model wasn't decorated enough.

As Java moves towards immutable objects like `record` the transformation is bottom up so the pain is even greater. 

It is a waste of time (and in Java object allocation) manually adding decoration such as index information so it is good to hear Mustache.java has a way to do this automatically. I was thinking it probably did but I haven't played with it in some time. Perhaps later today I will check it out again.

**Reply by spullara** [reply]
There are a lot of examples of various extensions in the tests which I use to show folks who ask "how do i do ...?"

example: https://github.com/spullara/mustache.java/blob/39dbf470d3eee90160c18cfba28e61b225d307be/compiler/src/test/java/com/github/mustachejava/util/DecoratedCollectionTest.java#L33

**Reply by jgonggrijp** [reply]
> Yeah my point to highlight how we do i18n is to show how varied this is.

You certainly opened my eyes!

> i18n is going to vary and that i18next seems to be @jgonggrijp major use case for power lambdas at the moment. I'm sure he has others and that is kind of what I want to see.

I'm afraid I don't.

I mean, I used to. I scattered code examples across #135 and a few other tickets on how power lambdas could do all kinds of wonderful things: list enumeration with first/last detection, list nonemptiness checks, key/value iteration over hashes, stack unmasking, dynamic partials, i18n and possibly a few others. However, I no longer endorse any of those examples. Even the i18n approach I illustrated above is very different from how I thought about it before.

I find it helpful to distinguish five flavors of lambdas (suggestions for better names welcome):

- "base lambdas", which return a value for interpolation or as the current context
- "extension lambdas", which return a template (that overrides a section if used there)
- "introspective lambdas", which have access to the contents of the stack while acting as a base or extension lambda
- "liberated lambdas", which can act as a base lambda and an extension lambda *at the same time* and which have more control over the way in which templates are overridden
- "full-on power lambdas", which are introspective and liberated at the same time

Introspective lambdas correspond to features 1-3, liberated lambdas correspond to features 4-7, and full-on power lambdas have all features.

When I first coined "power lambdas", I meant introspective lambdas. All motivating use cases needed stack access. My concept of "power lambdas" then gradually evolved into the full-on variant, as I realized that my i18n use case needed more than just stack access. This move was also inspired by @bobthecow mentioning feature 6 to me.

When first learning how filters worked (*also* due to @bobthecow), I recognized that they handled some of my use cases better than lambdas, but I still thought full-on power lambdas and filters could exist next to each other. More recently, however, I have come to realize that there is no added value in the introspective side of lambdas if we can have filters, in large part inspired by comments and critical questions by @agentgt. Lambdas are functions that can access the template, filters are functions that can access the stack, and together they can do everything.

Thus, my interest in power lambdas has eroded down to liberated lambdas. i18n is the only use case I can think of right now that I really need them for. I now realize that a majority of Mustache users may not need power lambdas at all.

To conclude, I guess they don't belong in the spec? Maybe it is more of a soft recommendation kind of thing, as was also suggested for ambient context? Or maybe not even that?

**Reply by agentgt** [reply]
Just to play around with your i18n example in JStachio I almost got the following to work:

```java
public interface I18NPowerFilter {

    @JStacheLambda
    default I18N i18n(Object payload) {
        return new I18N(payload);
    }

    public class I18N {

        private final Object payload;

        private List<String> keys = new ArrayList<>();

        public I18N(Object payload) {
            this.payload = payload;
        }

        @JStacheLambda
        public @JStacheLambda.Raw String key(@JStacheLambda.Raw String content) {
            keys.add(content);
            return ""; // we can't do null as I think it breaks stuff
        }

        @JStacheLambda
        public @JStacheLambda.Raw String t(@JStacheLambda.Raw String content) {
            // var translation = i18next.t(this.keys, content, payload);
            // we pretend to do the translation here
            var translation = "TRANSLATED keys:" + keys + " content: '" + content + "' payload: '" + payload + "'";
            // probably some smart caching would be done here
            // We use JMustache here since jstachio cannot handle dynamic templates
            return Mustache.compiler().compile(translation).execute(payload);

        }

    }

}

record Friend(String name) {
}

@JStache(template = "{{# friend.i18n }}{{#key}}goodmorning{{/key}}{{#t}}Good morning, {{name}}{{/t}}{{/ friend.i18n }}")
record PowerFilterExample(Friend friend) implements I18NPowerFilter {
}
```

It should work but lambda returning lambda is broken. There is no technical reason it should not other than the bug.

**Reply by spullara** [reply]
I think this just works in mustache.java. Like most things I can usually figure out a way to do it presuming this has the semantics you are expecting:

https://github.com/spullara/mustache.java/blob/main/compiler/src/test/java/com/github/mustachejava/SpecDisc155Test.java

**Reply by agentgt** [reply]
@spullara I think the payload is supposed to be passed to some translation piece (what it would know what to do with it I'm not sure)? Otherwise I would agree based on a cursory look your code is largely equivalent.

**Reply by spullara** [reply]
That would be the part where it does a lookup in the translation hash map of the translation key as a stand in for some real translation system with a fallback to the default text provided. As a bonus this also demonstrates having a global context that does the I18N work.

**Reply by agentgt** [reply]
I guess I'm just not sure what happens here given that payload is an object:

```js
var translation = i18next.t(this.keys, content, payload);
```

The `payload` in the above is the `friend` object (unlikely to be a key). Maybe it was just a typo: https://www.i18next.com/overview/api#t

**Reply by jgonggrijp** [reply]
@spullara your example seems to rest on the assumption that `i18next.t(['goodmorning'], 'Good morning, {{name}}', {name: 'Sam'})` is supposed to return `'좋은 아침이에요 {{name}}'`. It is not; it returns `'좋은 아침이에요 Sam'`. Also, your `t` lambda is not accessing the `payload` from my example, and I see no way in which it could. @agentgt recognized that correctly. Nice touch with the Korean, by the way.

@agentgt The i18next documentation is admittedly less than crystal clear. The `payload` argument is explained in these documentation sections:

- https://www.i18next.com/translation-function/essentials#overview-options (see `replace`)
- https://www.i18next.com/translation-function/interpolation#basic

**Reply by spullara** [reply]
Gotcha. Didn't realize that i18next is also a templating language that differs enough from mustache you can't figure out what needs to be made available. So are you planning on passing the entire context as JSON? That isn't possible in mustache.java since what fields that are available aren't enumerable. If it was enumerable it would be pretty easy to support it, but it is instead impossible unless you limit what is allowed in the context to Maps or JSON or some other pure data objects. IF you limited i18next content to compatible mustache templates I could build it as well OR if there was a way to ask i18next what keys are accessed by the content.

Mixing templating languages like this seems like a bad idea.

**Reply by jgonggrijp** [reply]
Right, so these are my two points:

1. Mixing templating languages is one of the things I need in this case, and I can do it in a safe way by the control inversion that would be unlocked by power lambdas (liberated flavor).
2. Retrieving arbitrary information from the context stack is the other thing I need. In my example, I used a filter to do it, but the introspective flavor of power lambdas would also do the job.

I don't need to pass the entire stack to i18next, I can just select a single frame because I know the default string. i18next is not really a templating language. It only does interpolation, but it does it in a localization-aware way that I think would be beyond the scope of Mustache.
